### PR TITLE
python3Packages.pydantic-core: init at 2.3.0

### PR DESCRIPTION
**Pull request by blaggacao**
As suggested by @SuperSandro2000, I'm targeting `staging` due to the dependency
on [a fix to the `maturinBuildHook`](https://github.com/NixOS/nixpkgs/pull/243859) which triggers a mass rebuild of ~1800+ packages.

Old PR to `master`: https://github.com/NixOS/nixpkgs/pull/242543

###### Description of changes

This package provides the core functionality for pydantic V2 validation and serialization.
Pydantic-core is currently around 17x faster than pydantic V1.

When this package is reviewed, attested and merged, we can start considering to bump pydantic to V2.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/pydantic-core/default.nix [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    typing-extensions
+    pythonRelaxDepsHook
+  ];
+
+  propagatedBuildInputs = [
+    typing-extensions
+  ];
+
+  pythonRelaxDeps = [
+    # NOTE: bumping typing-extensions to 4.7.1 would
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pydantic-core/default.nix:47]: We usually don't comment those things at all

File: pkgs/development/python-modules/pydantic-core/default.nix [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    # TODO: remove when typing-extensions is > 4.7.0
+    "typing-extensions"
+  ];
+
+  pythonImportsCheck = [ "pydantic_core" ];
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    hypothesis
+    pytest-benchmark
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pydantic-core/default.nix:59]:
```suggestion
```
We don't want to use compute power for benchmarking

**Comment by SuperSandro2000** [comment]
Please undraft when the other PR is merged

File: pkgs/development/python-modules/pydantic-core/default.nix (hunk 2 above)
Comment (blaggacao) [reply pkgs/development/python-modules/pydantic-core/default.nix:59]: It is disabled in `disabledTestPaths`, but still needs to be passed due to [`pyproject.toml` flags](https://github.com/pydantic/pydantic-core/blob/9be711e35760c248094367e0aa117ea7b3013909/pyproject.toml#L76-L81) that are only available via this plugin.

File: pkgs/development/python-modules/pydantic-core/default.nix (hunk 1 above)
Comment (blaggacao) [reply pkgs/development/python-modules/pydantic-core/default.nix]: Removed in [force-push](https://github.com/NixOS/nixpkgs/compare/951c4557822991b5c27b5d4819a091b9b956618a..4902a0946c02feef94b1a3b5578a13a611925694)

File: pkgs/development/python-modules/pydantic-core/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/pydantic-core/default.nix:55]: Just patch them out to not trigger unnecessary rebuilds

Comment (blaggacao) [reply pkgs/development/python-modules/pydantic-core/default.nix:55]: Oh yeah, good point! Thx!

Comment (blaggacao) [reply pkgs/development/python-modules/pydantic-core/default.nix]: Done so in [force-push](https://github.com/NixOS/nixpkgs/compare/e6488e846cee733f352ed57c2415b9e5a1bf25db..173deded1e38c279baabe88a448e657ffc595edd). Also removed the typing-extensions relaxation after [the update has been merged into staging](https://github.com/NixOS/nixpkgs/pull/243861), already.

**Comment by SuperSandro2000** [comment]
@ofborg build python310Packages.pydantic-core

**Pull request merged**